h/static/scripts/streamer.js

```diff
@@ -148,12 +98,78 @@ function connect($rootScope, annotationMapper, groups, session, settings) {
         console.warn('received unsupported notification', message.type);
       }
     });
-  });
+  }
 
-  return controls;
+  function sendClientConfig () {
+    Object.keys(configMessages).forEach(function (key) {
+      if (configMessages[key]) {
+        socket.send(configMessages[key]);
+      }
+    });
+  }
+
+  /**
+   * Send a configuration message to the push notification service.
+   * Each message is associated with a key, which is used to re-send
+   * configuration data to the server in the event of a reconnection.
+   */
+  function setConfig(key, configMessage) {
+    configMessages[key] = configMessage;
+    if (socket && socket.isConnected()) {
+      socket.send(configMessage);
+    }
+  }
+
+  var connect = function connect () {
+    var self = this;
+    var sidebarOpen = false;
+    var url = settings.websocketUrl;
+
+    // If we have no URL configured, don't do anything.
+    if (!url) {
+      return;
+    }
+
+    // Close any existing socket
+    if (socket) {
+      socket.close();
+    }
+
+    socket = new Socket(url);
+
+    socket.on('open', sendClientConfig);
+    socket.on('error', handleSocketOnError);
+    socket.on('message', handleSocketOnMessage);
+
+    $rootScope.$on('sidebarOpened', function () {
```

Comment:
I know this was taken from my initial patch, but I think it would be better to hoist the handling of the sidebar opening and the username changing here to `WidgetController` alongside the other logic. That way the streamer doesn't need to know about the sidebar state and it should be easier to add tests for this handling in `widget-controller-test.js`.